lib/src/winrt/foundation/collections/ikeyvaluepair.dart

```diff
@@ -0,0 +1,278 @@
+// ikeyvaluepair.dart
+
+// ignore_for_file: constant_identifier_names, non_constant_identifier_names
+
+import 'dart:ffi';
+
+import 'package:ffi/ffi.dart';
+
+import '../../../api_ms_win_core_winrt_string_l1_1_0.dart';
+import '../../../com/iinspectable.dart';
+import '../../../combase.dart';
+import '../../../exceptions.dart';
+import '../../../guid.dart';
+import '../../../macros.dart';
+import '../../../types.dart';
+import '../../../utils.dart';
+import '../../../winrt_helpers.dart';
+import '../../devices/sensors/enums.g.dart';
+import '../../internal/map_helpers.dart';
+import '../ipropertyvalue.dart';
+
+/// Represents a key-value pair.
+///
+/// This is typically used as a constraint type when you need to encapsulate
+/// two type parameters into one to satisfy the constraints of another generic
+/// interface.
+///
+/// {@category Interface}
+/// {@category winrt}
+class IKeyValuePair<K, V> extends IInspectable {
+  // vtable begins at 6, is 2 entries long.
+  final V Function(Pointer<COMObject>)? _creator;
+  final V Function(int)? _enumCreator;
+
+  /// Creates an instance of [IKeyValuePair] using the given [ptr].
+  ///
+  /// [K] must be of type `GUID`, `int`, `Object`, `String`, or `WinRTEnum`
+  /// (e.g. `PedometerStepKind`).
+  ///
+  /// [V] must be of type `Object`, `String`, or `WinRT` (e.g. `IJsonValue`,
+  /// `ProductLicense`).
+  ///
+  /// [creator] must be specified if [V] is a `WinRT` type.
+  /// ```dart
+  /// final keyValuePair =
+  ///     IKeyValuePair<String, IJsonValue?>.fromRawPointer(ptr,
+  ///         creator: IJsonValue.fromRawPointer);
+  /// ```
+  ///
+  /// [enumCreator] must be specified if [V] is a `WinRTEnum` type.
+  /// ```dart
+  /// final keyValuePair =
+  ///     IKeyValuePair<String, ChatMessageStatus?>.fromRawPointer(ptr,
+  ///         enumCreator: ChatMessageStatus.from);
+  /// ```
+  IKeyValuePair.fromRawPointer(
+    super.ptr, {
+    V Function(Pointer<COMObject>)? creator,
+    V Function(int)? enumCreator,
+  })  : _creator = creator,
+        _enumCreator = enumCreator {
+    if (!isSupportedKeyValuePair<K, V>()) {
+      throw ArgumentError('Unsupported key-value pair: IKeyValuePair<$K, $V>');
+    }
+
+    if (isSubtypeOfInspectable<V>() && creator == null) {
+      throw ArgumentError.notNull('creator');
+    }
+
+    if (isSubtypeOfWinRTEnum<V>() && enumCreator == null) {
+      throw ArgumentError.notNull('enumCreator');
+    }
+  }
+
+  /// Gets the key of the key-value pair.
+  K get key {
+    if (isSameType<K, GUID>()) return _key_GUID as K;
+    if (isSameType<K, int>()) return _key_Uint32 as K;
```

Comment:
I used `Uint32` for `int` keys since it is the only `int` type to be supported for `IKeyValuePair`: https://github.com/timsneath/win32/blob/339c6e4ac04a9d2f2c1933c958bc0893be65dee2/lib/src/winrt_constants.dart#L147-L148